### DISCUSSION TITLE
So would it be possible to implement sending `SET ROLE` based on some predefined JWT attributes?

### DISCUSSION THREAD

**Discussion by avelino** [prest/prest#549] [API Server]
So would it be possible to implement sending `SET ROLE` based on some predefined JWT attributes? Any reason why it should not work? Any coding tips in case I want to try to code it my self?

We need it for the project I am working on now. The alternative is to throw out pREST and make our own API.

_Originally posted by @hjelmeland in https://github.com/prest/prest/issues/197#issuecomment-858466782_

**Comment by avelino** [comment]
@hjelmeland have you seen the new authentication system recently implemented in pREST? https://docs.prestd.com/configurations/#auth
With it you can set up a table, scheme (login and password field) and cryptorphy used in the password.

Another possibility would be to use [prest as a framework](https://docs.prestd.com/prest-as-web-framework/) and create a custom middleware, see [here an example](https://github.com/prest/prest-example/blob/e8d9984ecffb4d964183e194f1955b32ff9b8781/custom-middlewares/main.go#L24)

**Reply by hjelmeland** [reply]
Yes I have, but will that make it possible to grant different table access permissions per user?

**Reply by avelino** [reply]
by the authentication system no, via middleware (prest as framework) yes

As I understand you already use prest and using it as a framework will give you less work than developing new software (API)

**Reply by hjelmeland** [reply]
Thanks for swift response! I have to look into this middleware thing then. Have the middleware layer access to the database connection? Could a middleware function send "RESET ROLE; SET ROLE xxx;" to postgres before the table controllers are called? I understand like there is just one connection per database...

**Reply by hjelmeland** [reply]
We are using RSA signed JWTs, so I solved it with
```diff
-- a/middlewares/middlewares.go
+++ b/middlewares/middlewares.go
@@ -106,6 +106,9 @@ func AccessControl() negroni.Handler {
 func JwtMiddleware(key string, algo string) negroni.Handler {
        jwtMiddleware := jwtmiddleware.New(jwtmiddleware.Options{
                ValidationKeyGetter: func(token *jwt.Token) (interface{}, error) {
+                       if (algo[0:2] == "RS") {
+                               return jwt.ParseRSAPublicKeyFromPEM( []byte(key))
+                       }
                        return []byte(key), nil
                },
                SigningMethod: jwt.GetSigningMethod(algo),

```
But it would been better if ParseRSAPublicKeyFromPEM() was called once in stead of for each request. Suggestions?

**Reply by hjelmeland** [reply]
Ah, I think I got it 

```diff
--- a/middlewares/middlewares.go
+++ b/middlewares/middlewares.go
@@ -104,12 +104,24 @@ func AccessControl() negroni.Handler {
 
 // JwtMiddleware check if actual request have JWT
 func JwtMiddleware(key string, algo string) negroni.Handler {
-       jwtMiddleware := jwtmiddleware.New(jwtmiddleware.Options{
-               ValidationKeyGetter: func(token *jwt.Token) (interface{}, error) {
-                       return []byte(key), nil
-               },
-               SigningMethod: jwt.GetSigningMethod(algo),
-       })
+       var jwtMiddleware *jwtmiddleware.JWTMiddleware
+
+       if (algo[0:2] == "RS") {
+               rsaKey, err  := jwt.ParseRSAPublicKeyFromPEM( []byte(key))
+               jwtMiddleware = jwtmiddleware.New(jwtmiddleware.Options{
+                       ValidationKeyGetter: func(token *jwt.Token) (interface{}, error) {
+                               return rsaKey, err
+                       },
+                       SigningMethod: jwt.GetSigningMethod(algo),
+               })
+       } else {
+               jwtMiddleware = jwtmiddleware.New(jwtmiddleware.Options{
+                       ValidationKeyGetter: func(token *jwt.Token) (interface{}, error) {
+                               return []byte(key), nil
+                       },
+                       SigningMethod: jwt.GetSigningMethod(algo),
+               })
+       }
 ```
